### PR TITLE
Fix #3738 createBuilding optional arguments not handled correctly

### LINKED ISSUE #3738: createBuilding optional arguments not handled correctly
### Describe the bug

doing code below will throw error for arg 5 expect vector get none

### Steps to reproduce

`createBuilding ( 3534, -2412.30884, -601.68127, 132.56250)`
or
`createBuilding ( 3534, -2412.30884, -601.68127, 132.56250,90)` without rz or ry gives error

### Version

_No response_

### Additional context

_No response_

### Relevant log output

_No response_

### Security Policy

- [X] I have read and understood the Security Policy and this issue is not security related.

### PR DESCRIPTION
**Pull request by FileEX**
Fix #3738 

I think this is what the parser should take care of, that if the vector is optional, it should put 0 in the remaining fields

### CODE REVIEW

File: Client/mods/deathmatch/logic/luadefs/CLuaBuildingDefs.cpp [hunk 1]
```diff
@@ -48,14 +48,13 @@ CClientBuilding* CLuaBuildingDefs::CreateBuilding(lua_State* const luaVM, std::u
     if (!CClientBuildingManager::IsValidPosition(pos))
         throw std::invalid_argument("Position is outside of game world");
 
-    if (rot.has_value())
-        ConvertDegreesToRadians(rot.value());
-    else
-        rot.emplace(CVector(0, 0, 0));
+    CVector rot{rotX.value_or(0), rotY.value_or(0), rotZ.value_or(0)};
+    if (rot.fX != 0 || rot.fY != 0 || rot.fZ != 0)
+        ConvertDegreesToRadians(rot);
```
Comment (TracerDS) [inline Client/mods/deathmatch/logic/luadefs/CLuaBuildingDefs.cpp:54]: wouldnt it be better to always convert to radians?

**Review by TracerDS — APPROVED** [review]

File: Client/mods/deathmatch/logic/luadefs/CLuaBuildingDefs.cpp (hunk 1 above)
Comment (FileEX) [reply Client/mods/deathmatch/logic/luadefs/CLuaBuildingDefs.cpp:54]: 0 deg = 0 rad. This way, unnecessary calculations can be avoided, just like the previous code did

Comment (tederis) [inline Client/mods/deathmatch/logic/luadefs/CLuaBuildingDefs.cpp]: There should be floating point literals (.0f). In addition to that `rot.fX != 0` can be replaced with `std::abs(rot.fX) < std::numeric_limits<float>::epsilon()` to make the comparison less strong.

**Comment by TheNormalnij** [comment]
I don't think that original issue is valid. Developers should provide all vector arguments. It has more sense for me.

You fixes only one case of this *issue*.  The main problem hides in argument parser.

**Comment by TracerDS** [comment]
> I don't think that original issue is valid. Developers should provide all vector arguments. It has more sense for me.
> 
> You fixes only one case of this _issue_. The main problem hides in argument parser.

Its more of a logic issue. What do you expect vector to receive? 3 coordinates.
What happens when you dont provide all 3 coordinates? Error occurs.

**Comment by FileEX** [comment]
TheNormalnij is right. This should be fixed in the parser. Many functions rely on the syntax where each argument in [ ] is optional, such as createVehicle, createObject, etc. The method in the old parser, ReadVector3D, would simply insert 0 in place of any missing argument. The new parser should handle this as well, because the current behavior is confusing. In one function, you can provide a single argument, but in another, you have to provide all three, even though each one is marked as optional individually

**Comment by TracerDS** [comment]
> TheNormalnij is right. This should be fixed in the parser. Many functions rely on the syntax where each argument in [ ] is optional, such as createVehicle, createObject, etc. The method in the old parser, ReadVector3D, would simply insert 0 in place of any missing argument. The new parser should handle this as well, because the current behavior is confusing. In one function, you can provide a single argument, but in another, you have to provide all three, even though each one is marked as optional individually

for now lets just treat vectors as `struct { float x,y,z; }`

File: Client/mods/deathmatch/logic/luadefs/CLuaBuildingDefs.cpp (hunk 1 above)
Comment (FileEX) [reply Client/mods/deathmatch/logic/luadefs/CLuaBuildingDefs.cpp]: Done

**Comment by Pirulax** [comment]
Should be an easy fix

> TheNormalnij is right. This should be fixed in the parser. Many functions rely on the syntax where each argument in [ ] is optional, such as createVehicle, createObject, etc. The method in the old parser, ReadVector3D, would simply insert 0 in place of any missing argument. The new parser should handle this as well, because the current behavior is confusing. In one function, you can provide a single argument, but in another, you have to provide all three, even though each one is marked as optional individually

So, just default initialize values (CVector for example)?

**Comment by FileEX** [comment]
> Should be an easy fix
> 
> > TheNormalnij is right. This should be fixed in the parser. Many functions rely on the syntax where each argument in [ ] is optional, such as createVehicle, createObject, etc. The method in the old parser, ReadVector3D, would simply insert 0 in place of any missing argument. The new parser should handle this as well, because the current behavior is confusing. In one function, you can provide a single argument, but in another, you have to provide all three, even though each one is marked as optional individually
> 
> So, just default initialize values (CVector for example)?

The point is that if a function accepts ``std::optional<CVector>``, we have to provide the entire vector, all three values, because otherwise, there's an error that says a vector was expected, but a number was received. It should work like in the old parser, where if, for example, two values ``fX`` and ``fY`` were provided, the ``fZ`` value would simply be set to 0 and treated as a valid vector

**Comment by Pirulax** [comment]
yeah thats what i meant
that should be easy

**Pull request closed without merging**